core/pkg/store/store.go

```diff
@@ -50,49 +52,79 @@ func (f *State) hasPriority(stored string, new string) bool {
 	return true
 }
 
-func NewFlags() *State {
-	return &State{
-		Flags:             map[string]model.Flag{},
+func NewStore(logger *logger.Logger) (*Store, error) {
+
+	schema := &memdb.DBSchema{
+		Tables: map[string]*memdb.TableSchema{
+			"flags": {
+				Name: "flags",
+				Indexes: map[string]*memdb.IndexSchema{
+
+					"id": {
+						Name:    "id",
+						Unique:  true,
+						Indexer: &memdb.StringFieldIndex{Field: "Key", Lowercase: false},
+					},
+				},
+			},
+		},
+	}
+
+	// Create a new data base
+	db, err := memdb.NewMemDB(schema)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize flag database: %w", err)
+	}
+
+	return &Store{
 		SourceDetails:     map[string]SourceDetails{},
 		MetadataPerSource: map[string]model.Metadata{},
-	}
+		db:                db,
+		logger:            logger,
+	}, nil
 }
 
-func (f *State) Set(key string, flag model.Flag) {
-	f.mx.Lock()
-	defer f.mx.Unlock()
-	f.Flags[key] = flag
+// Deprecated: use NewStore instead
+func NewFlags() *Store {
+	state, err := NewStore(logger.NewLogger(nil, false))
+	if err != nil {
+		panic(fmt.Sprintf("unable to create flag store: %v", err))
+	}
+	return state
 }
 
-func (f *State) Get(_ context.Context, key string) (model.Flag, model.Metadata, bool) {
+func (f *Store) Get(_ context.Context, key string) (model.Flag, model.Metadata, bool) {
+	f.logger.Debug(fmt.Sprintf("getting flag %s", key))
 	f.mx.RLock()
```

Comment:
We could move the lock operations to `GetMetadataForSource`.

Reply:
Done